src/Symfony/Component/HttpKernel/Controller/ArgumentResolver.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\ControllerMetadata\Argument\ArgumentMetadataFactoryInterface;
+
+/**
+ * Responsible for the resolving of arguments passed to an action.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+final class ArgumentResolver implements ArgumentResolverInterface
+{
+    /**
+     * @var ArgumentMetadataFactoryInterface
+     */
+    private $argumentMetadataFactory;
+
+    /**
+     * @var ArgumentValueResolverInterface[]
+     */
+    private $argumentValueResolvers;
+
+    /**
+     * @param ArgumentMetadataFactoryInterface $argumentMetadataFactory
+     * @param ArgumentValueResolverInterface[] $argumentValueResolvers
+     */
+    public function __construct(ArgumentMetadataFactoryInterface $argumentMetadataFactory = null, array $argumentValueResolvers = array())
+    {
+        $this->argumentMetadataFactory = $argumentMetadataFactory;
+        $this->argumentValueResolvers = $argumentValueResolvers;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getArguments(Request $request, $controller)
+    {
+        $arguments = array();
+
+        foreach ($this->argumentMetadataFactory->createArgumentMetadata($controller) as $metadata) {
+            $isResolved = false;
+            foreach ($this->argumentValueResolvers as $resolver) {
+                if ($resolver->supports($request, $metadata)) {
+                    $resolved = $resolver->getValue($request, $metadata);
+
+                    // variadic is a special case, always being the last and being an array
+                    if (is_array($resolved) && $metadata->isVariadic()) {
+                        return array_merge($arguments, $resolved);
+                    }
```

Comment:
Let's keep DX aside for a minute, not everything is DX; adding new resolvers is going to be really rare anyway.
Instead of `array_merge()`, `+` can be used.
Your alternative implementation is wrong as the returned value can be an array, so you won't be able to determine whether the value is an array or comes from a variadic.